playground/frontend/playground_components/lib/src/controllers/code_runner.dart

```diff
@@ -27,121 +27,98 @@ import 'snippet_editing_controller.dart';
+  String get resultOutput => _result?.output ?? '';
+  String get resultLogOutput => resultLog + resultOutput;
+
   void clearResult() {
     _result = null;
     notifyListeners();
   }
 
   void runCode({void Function()? onFinish}) {
     _runStartDate = DateTime.now();
+    _runStopDate = null;
     notifyListeners();
+    snippetEditingController = _snippetEditingControllerGetter();
 
     final parsedPipelineOptions =
-        parsePipelineOptions(_snippetEditingController().pipelineOptions);
+        parsePipelineOptions(snippetEditingController!.pipelineOptions);
     if (parsedPipelineOptions == null) {
       _result = const RunCodeResult(
         status: RunCodeStatus.compileError,
         errorMessage: kPipelineOptionsParseError,
       );
+      _runStopDate = DateTime.now();
       notifyListeners();
       return;
     }
 
-    if (!_snippetEditingController().isChanged &&
-        _snippetEditingController().selectedExample?.outputs != null) {
+    if (!snippetEditingController!.isChanged &&
+        snippetEditingController!.selectedExample?.outputs != null) {
       unawaited(_showPrecompiledResult());
     } else {
       final request = RunCodeRequest(
-        code: _snippetEditingController().codeController.fullText,
-        sdk: _snippetEditingController().sdk,
+        code: snippetEditingController!.codeController.fullText,
+        sdk: snippetEditingController!.sdk,
         pipelineOptions: parsedPipelineOptions,
       );
       _runSubscription = _codeRepository?.runCode(request).listen((event) {
         _result = event;
-        filterOutput(selectedOutputFilterType);
-
-        if (event.isFinished && onFinish != null) {
-          onFinish();
+        notifyListeners();
+
+        if (event.isFinished) {
+          if (onFinish != null) {
+            onFinish();
+          }
+          snippetEditingController = null;
+          _runStopDate = DateTime.now();
         }
       });
       notifyListeners();
     }
   }
 
-  void filterOutput(OutputType type) {
-    final output = _result?.output ?? '';
-    final log = _result?.log ?? '';
-
-    switch (type) {
-      case OutputType.all:
-        setOutputResult(log + output);
-        break;
-      case OutputType.log:
-        setOutputResult(log);
-        break;
-      case OutputType.output:
-        setOutputResult(output);
-        break;
-      case OutputType.graph:
-        setOutputResult(log + output);
-        break;
-    }
-  }
-
-  void setOutputResult(String outputs) {
-    outputResult = outputs;
-    notifyListeners();
-  }
-
   void clearOutput() {
     _result = null;
     notifyListeners();
   }
 
-  void setSelectedOutputFilterType(OutputType type) {
-    selectedOutputFilterType = type;
-    notifyListeners();
-  }
-
-  void reset() {
-    _snippetEditingController().reset();
-    outputResult = '';
-    notifyListeners();
-  }
-
-  void resetError() {
-    if (result == null) {
+  void setResultAfterResetError() {
```

Comment:
```suggestion
  /// Resets the error message text so that on the next rebuild
  /// of `CodeTextAreaWrapper` it is not picked up and not shown as a toast.
  // TODO: Listen to this object outside of widgets,
  //  emit toasts from notifications, then remove this method.
  void resetErrorMessageText() {
```